### PR TITLE
Do TDNR substitutions in a single term traversal

### PR DESCRIPTION
**Pull request by ChrisPenner**
## Overview

Just noticed this when familiarizing myself with the parser-typechecking codepath;

looks like TDNR substitution is performing a full ABT visit for each TDNR 'decision', I think we can do them all in a single pass by performing decision lookups instead, which should give us a meagre improvement.

If we have a term of size `t` has `d` TDNR decisions, this goes from `O(t * d)` down to `O(t * log(d))`, the `log(d)` represents the map lookup.

## Questions:

* Is there a reason we might need these to be sequential? AFAICT the only thing that might cause an issue here is if replacements themselves had Blanks in them, but I'm pretty sure they never do.

## Implementation notes

* Performs TDNR in a single visit over the term rather than folding over TDNR decisions.
* Converts the `doTDNR` function to be pure, since it doesn't use its effects.

## Testing

Not sure how I'd test this specifically, I suspect our existing transcripts cover a decent amount of TDNR cases already.

- - -
This change is [Reviewable](https://reviewable.io/reviews/unisonweb/unison/2789)

### CODE REVIEW

File: parser-typechecker/src/Unison/FileParsers.hs [hunk 1]
```diff
@@ -176,9 +176,8 @@ synthesizeFile ambient tl fqnsByShortName uf term = do
         -- use tlcsFromTypechecker to inform annotation-stripping decisions
         traverse (traverse strippedTopLevelBinding) tlcsFromTypechecker
     let doTdnr = applyTdnrDecisions infos
-        doTdnrInComponent (v, t, tp) = (\t -> (v, t, tp)) <$> doTdnr t
-    _          <- doTdnr tdnrTerm
```
Comment (ChrisPenner) [inline parser-typechecker/src/Unison/FileParsers.hs]: `doTdnr` is pure now (and didn't use any effects before), so this seems to have been doing nothing;
Anyone know what the original purpose of this was?

File: parser-typechecker/src/Unison/FileParsers.hs [hunk 2]
```diff
@@ -198,19 +197,15 @@ synthesizeFile ambient tl fqnsByShortName uf term = do
   applyTdnrDecisions
     :: [Context.InfoNote v Ann]
     -> Term v
-    -> Result' v (Term v)
```
Comment (ChrisPenner) [inline parser-typechecker/src/Unison/FileParsers.hs]: The implementation always succeeds with `pure` and never fails, so this function is actually pure.

**Comment by aryairani** [comment]
I didn't look too closely at the diff, but I think there had been a thing where it could be that it might not be possible to solve all types simultaneously, but could be possible to solve them one at a time, if each single substitution provided enough of a constraint to solve the next.  Would that explain what you found? /cc @runarorama

**Comment by runarorama** [comment]
That sounds right. The assumption is that there won't be many terms to look up by type, but maybe that's wrong.

File: parser-typechecker/src/Unison/FileParsers.hs (hunk 1 above)
Comment (aryairani) [reply parser-typechecker/src/Unison/FileParsers.hs]: I'm not too familiar with this code, but you could check with @runarorama or try to dig it out of git history.

**Comment by aryairani** [comment]
I guess we should add a regression test for this behavior if we want to preserve it.

**Comment by aryairani** [comment]
@ChrisPenner I moved this PR into the "New" category so that we can triage it.

**Comment by pchiusano** [comment]
For better or worse, TDNR proceeds in rounds, where initially, all the TDNR variables are unsolved, and after each round we get a list of substitutions. These get substituted into the term and there's another round. This continues until all the variables are solved or a round produces no new substitutions.

If this is just optimizing the application of TDNR decisions from a single round, it seems fine. If it's reducing the number of TDNR rounds to 1, then it's a change in behavior which will reduce what TDNR can do. Though this might also be fine and might even be a more predictable if more limited algorithm? But that would need more thought.

**Comment by aryairani** [comment]
@pchiusano Yeah, we think this is the former (just optimizing a single round), but let's add a regression test for the multi-round behavior we want and then go from there.

If we think we're demoing right behavior, and the PR still passes, then let's merge it.

**Comment by aryairani** [comment]
@ChrisPenner Let's pair today or in the next few days to try to knock this out.

**Comment by aryairani** [comment]
It would be nice if Github's "Resolve Conflicts" did a 3-way merge.

**Review by aryairani — APPROVED** [review]

**Pull request merged**